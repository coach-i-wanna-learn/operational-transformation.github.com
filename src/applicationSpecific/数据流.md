
```mermaid
sequenceDiagram
CodeMirrorAdapter ->> MirrorComponent: replaceRange
MirrorComponent ->> CodeMirrorAdapter: operationFromCodeMirrorChanges
MirrorComponent ->> Visualization: onUserChange
MirrorComponent ->> Visualization: onClientOperation
%%  这里传入 state，onClientOperation onServerReceiveClick onClientReceiveClick
makeVisualization ->> makeClientAndSocketsVisualization: makeClientAndSocketsVisualization
makeClientAndSocketsVisualization ->> MirrorComponent: ref onUserChange=onClientOperation
```


```mermaid
sequenceDiagram
%% VisualizationState 包含 alice bob server
makeVisualization ->> makeVisualization: visualizationState
makeVisualization ->> ClientAndSocketsVisualization: visualizationState bob or alice
ClientAndSocketsVisualization ->> MirrorComponent: applyOperation 更新 codemirror 中的内容
```
