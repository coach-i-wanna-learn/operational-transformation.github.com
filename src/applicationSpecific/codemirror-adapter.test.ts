import { CodeMirrorAdapter } from '@/applicationSpecific/codemirror-adapter'

describe('CodeMirrorAdapter', () => {
  it('operationFromCodeMirrorChanges', () => {
    const testcases = [
      {
        input: {
          changes: [
            {
              from: { line: 0, ch: 13, sticky: null },
              to: { line: 0, ch: 13, sticky: null },
              text: ['1'],
              removed: [''],
              origin: '+input',
            } as any,
          ],
          doc: {
            indexFromPos: () => 0,
            getLine: () => 'Lorem ipsum1234',
            lastLine: () => 0,
          } as any,
        },
      },
      {
        input: {
          changes: [
            {
              from: { line: 0, ch: 13, sticky: null },
              to: { line: 0, ch: 13, sticky: null },
              text: ['11'],
              removed: [''],
              origin: '+input',
            } as any,
          ],
          doc: {
            indexFromPos: () => 0,
            getLine: () => 'Lorem ipsum1234',
            lastLine: () => 0,
          } as any,
        },
      },
      {
        input: {
          changes: [
            {
              from: { line: 0, ch: 1, sticky: null },
              to: { line: 0, ch: 1, sticky: null },
              text: ['1'],
              removed: [''],
              origin: '+input',
            } as any,
          ],
          doc: {
            indexFromPos: ({ line, ch }: { line: number; ch: number }) =>
              11 * line + line + ch,
            getLine: () => 'L1orem ipsum',
            lastLine: () => 0,
          } as any,
        },
      },
      {
        input: {
          changes: [
            {
              from: { line: 1, ch: 11, sticky: null },
              to: { line: 1, ch: 11, sticky: null },
              text: ['1'],
              removed: [''],
              origin: '+input',
            } as any,
          ],
          // 这里假定除了最后一行每一行都是 Lorem ipsum
          doc: {
            indexFromPos: ({ line, ch }: { line: number; ch: number }) =>
              11 * line + line + ch,
            getLine: () => 'Lorem ipsum1',
            lastLine: () => 1,
          } as any,
        },
      },
    ]
    testcases.forEach(({ input }) => {
      const [operation, inverse] =
        CodeMirrorAdapter.operationFromCodeMirrorChanges(
          input.changes,
          input.doc,
        )

      expect([operation.toString(), inverse.toString()]).toMatchSnapshot()
    })
  })
})
