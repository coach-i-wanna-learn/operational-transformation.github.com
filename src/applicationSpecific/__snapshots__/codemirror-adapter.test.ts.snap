// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeMirrorAdapter operationFromCodeMirrorChanges 1`] = `
[
  "insert '1', retain 14",
  "delete 1, retain 14",
]
`;

exports[`CodeMirrorAdapter operationFromCodeMirrorChanges 2`] = `
[
  "insert '11', retain 13",
  "delete 2, retain 13",
]
`;

exports[`CodeMirrorAdapter operationFromCodeMirrorChanges 3`] = `
[
  "retain 1, insert '1', retain 10",
  "retain 1, delete 1, retain 10",
]
`;

exports[`CodeMirrorAdapter operationFromCodeMirrorChanges 4`] = `
[
  "retain 23, insert '1'",
  "retain 23, delete 1",
]
`;
